### ISSUE TITLE
Cannot use left/right arrow keys to exit text in Console.ReadLine on MacOS

### ISSUE THREAD

**Issue by nickorlow** [dotnet/runtime#38051]
### Description

On Windows, using the left/right arrow keys lets you modify the middle of a set of text from Console.ReadLine without having to delete any of it. However, on MacOS this feature is missing, making it so that you have to delete part of the text in order to modify other parts of it.

### Configuration

Found with MacOS, not tested with Linux. Windows has this functionality already.

I intend to attempt to fix this.

**Comment by stephentoub** [comment]
https://github.com/dotnet/runtime/issues/15711

**Comment by carlossanlop** [comment]
@adamsitnik @Jozkee @stephentoub I'm thinking we should move this to 7.0 for the following reasons:

- PR https://github.com/dotnet/runtime/pull/49649 was looking to fix this. The original contributor could not continue working on it, but @eiriktsarpalis and @nickorlow took over and pushed some additional commits to the PR.
- Unfortunately, some OSX regressions were introduced with those last commits, which still need to be investigated and fixed.
- @tmds merged PR https://github.com/dotnet/runtime/pull/49649 a few days ago, which heavily modified code that this issue's PR is also touching.
- I am currently reviewing both PRs to understand how to merge them, but I'm noticing it's not trivial.
- The change affects all platforms, and manual testing is required to verify this fix.
- I am not sure we will have enough time to get feedback from the community to ensure we didn't introduce regressions.

Does anyone have a strong reason to merge this in 6.0, or can we move it to 7.0?

**Comment by danmoseley** [comment]
Makes sense to me. We have seen subtle regressions in this code in the past.

**Comment by adamsitnik** [comment]
@carlossanlop sounds good to me, thanks for trying!

**Comment by nickorlow** [comment]
Hey maintainers cc: @stephentoub  @jeffschwMSFT 

I figured I'd take another stab at this since it seems like nobody is working on it right now

**Comment by adamsitnik** [comment]
> I figured I'd take another stab at this since it seems like nobody is working on it right now

@nickorlow Awesome! Please let us know if you need any help!

**Comment by nickorlow** [comment]
Update on my fix: I've been able to re-implement my changes from when I did this the first time. I was also able to fix 2 of the regressions that it introduced last time. Currently there's a bug involving wrapping multiple lines that I haven't seen before.  Once that's fixed,  I think I should be ready for a PR.

**Comment by tmds** [comment]
I was hoping PowerShell `ReadLineFromConsole` would have a good solution for this and other improvements to ReadLine (https://github.com/dotnet/runtime/issues/94860), but after looking further into it I learned PowerShell actually uses  https://github.com/PowerShell/PSReadLine.

There is no perfect solution to this problem, so what we do needs to be a 'good enough'.

One of the challenging cases that should work is: typing wrapped lines at the bottom of the terminal.
I don't think terminals offer a way to detect this wrapping except by determining the console position after each key press.

Other challenges include:
- Ensure we preserve the prompt that may have been written on the line before the ReadLine call.
- Characters that span multiple colums, like emojis.
- Handing window resizes. Other apps don't seem to handle this perfect.

@nickorlow did you make more progress on your fix? How does it work in the above cases?
